nanomq/apps/broker.c

```diff
@@ -474,16 +476,146 @@ broker(const char *url)
 #endif
 }
 
-int
-broker_start(int argc, char **argv)
+void *
+print_usage(void)
 {
+	fprintf(stderr, USAGE);
+}
+
+int status_check(pid_t *pid)
+{
+	char* data = NULL;
+	size_t size = 0;
+
 	int rc;
-	if (argc != 1) {
-		fprintf(stderr, "Usage: broker start <url>\n");
+	if ((rc = nng_file_get(PID_PATH_NAME, (void*)&data, &size)) != 0) {
+		debug_msg(".pid file does not existed or cannot be read");
+		return 1;
+
+	} else {
+		if ((data) != NULL ) {
+			sscanf(data, "%lu", pid);
+			debug_msg("pid read, [%lu]", *pid);
+
+			if ((kill(*pid, 0)) == 0) {
+				debug_msg("there is a running NanoMQ instance has pid [%lu]", *pid);
+				return 0;
+			}
+		}
+		if (!nng_file_delete(PID_PATH_NAME)) {
+			debug_msg(".pid file successfully deleted");
+			return 1;
+		}
+		debug_msg("unexpected error");
+		return -1;
+	}
+}
+
+int store_pid()
+{
+	int status;
+	char pid_c[10] = "";
+
+	sprintf(pid_c, "%ld", getpid());
+	debug_msg("%s", pid_c);
+
+	status = nng_file_put(PID_PATH_NAME, pid_c, sizeof(pid_c));
+	return status;
+	
+}
+
+int broker_start(int argc, char **argv)
+{
+	int   i, url, temp, rc, num_ctx = 0;
+	pid_t pid = 0;
+
+	if (argc < 1) {
+		print_usage();
 		exit(EXIT_FAILURE);
 	}
-	rc = broker(argv[0]);
+
+	if (!status_check(&pid)) {
+		fprintf(stderr, "One NanoMQ instance is running, a new instance won't be started until the other one is stopped.\n");
+		exit(EXIT_FAILURE);
+	}
+
+	for (i = 0; i < argc; i++, temp = 0) {
```

Comment:
Error. var `i` should start from 3(nanomq broker start -daemon) and the var argc it not really argc but argc-2

Reply:
When entering this function, the `argc` and `argv` have already been decremented and incremented respectively. So, here it is good for `i` to start from 1, and for `argc` to keep as it is.